tsfresh/feature_extraction/data.py

```diff
@@ -0,0 +1,343 @@
+
+        if "__" in str(col):
+            raise ValueError("Dict keys are not allowed to contain '__': {}".format(col))
+
+
+def _check_nan(df, *columns):
+    """
+    Raise a ``ValueError`` if one of the columns does not exist or contains NaNs.
+
+    :param df: the pandas DataFrame to test for NaNs
+    :type df: pandas.DataFrame
+    :param columns: a list of columns to test for NaNs. If left empty, all columns of the DataFrame will be tested.
+    :type columns: str
+
+    :return: None
+    :rtype: None
+    :raise: ``ValueError`` if ``NaNs`` are found in the DataFrame.
+    """
+
+    for col in columns:
+        if col not in df.columns:
+            raise ValueError("Column not found: {}".format(col))
+
+        if df[col].isnull().any():
+            raise ValueError("Column must not contain NaN values: {}".format(col))
+
+
+class WideTsFrameAdapter(SliceableTsData):
+
+    def __init__(self, df, column_id, column_sort=None, value_columns=None):
+        """
+        Adapter for Pandas DataFrames in wide format, where multiple columns contain different time series for
+        the same id.
+
+        :param df: the data frame
+        :type df: pd.DataFrame
+
+        :param column_id: the name of the column containing time series group ids
+        :type column_id: str
+
+        :param column_sort: the name of the column to sort on
+        :type column_sort: str|None
+
+        :param value_columns: list of column names to treat as time series values.
+            If `None`, all columns except `column_id` and `column_sort` will be used.
+        :type value_columns: list[str]|None
+        """
+
+        _check_nan(df, column_id)
+
+        if value_columns is None:
+            value_columns = [col for col in df.columns if col not in [column_id, column_sort]]
+
+        if len(value_columns) == 0:
+            raise ValueError("You must provide at least one value column")
+
+        _check_nan(df, *value_columns)
+        _check_colname(*value_columns)
+
+        self.value_columns = value_columns
+        self.column_sort = column_sort
+
+        if self.column_sort is not None:
+            _check_nan(df, column_sort)
+            self.df_grouped = df.sort_values([self.column_sort]).groupby([column_id])
+        else:
+            self.df_grouped = df.groupby([column_id])
+
+    def __len__(self):
+        return self.df_grouped.ngroups * len(self.value_columns)
+
+    def slice(self, offset, length=None):
+        group_offset, column_offset = divmod(offset, len(self.value_columns))
+        kinds = self.value_columns[column_offset:]
+        i = 0
+
+        for group_name, group in itertools.islice(self.df_grouped, group_offset, None):
+            for kind in kinds:
+                i += 1
+                if length is not None and i > length:
+                    return
+                else:
+                    yield Timeseries(group_name, kind, group[kind])
+            kinds = self.value_columns
```

Comment:
Maybe I am blind, but what makes this function (together with `slice` and the `partition` from `SliceableTsData`) faster, than just implementing `__iter__` as:
```python
def __iter__(self):
    for group_name, group in itertools.islice(self.df_grouped, 0, None):
        for kind in kinds:
            yield Timeseries(group_name, kind, group[kind])
```
and not deriving from `SliceableTsData` at all.

The python function `itertools.islice` does roughly the same as your own implementation of `partition` in `SliceableTsData` - with the difference that you can request chunks of a given length at once (which could be faster), whereas `itertools.islice` needs to ask the generator for `next` until it has enough for a chunk.
However here, you do not use this feature and need to iterate through the full loop anyways (one step at the time).
Just want to understand :-)

Reply:
The idea is, that I can calculate the group offset and skip whole groups instead of iterating all elements (groups * columns) using `itertools.islice`.